extensions/amp-addthis/0.1/amp-addthis.js

```diff
@@ -0,0 +1,216 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Embeds an AddThis widget.
+ * The data-pub-id and data-widget-id can be found easily in the AddThis dashboard at addthis.com.
+ * Example:
+ * <code>
+ * <amp-addthis
+ *   width="320"
+ *   height="92"
+ *   layout="responsive"
+ *   data-pub-id="ra-59c2c366435ef478"
+ *   data-widget-id="0fyg">
+ * </amp-addthis>
+ * </code>
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {setStyle} from '../../../src/style';
+import {createElementWithAttributes, removeElement} from '../../../src/dom';
+import {user} from '../../../src/log';
+
+import {ConfigManager} from './ConfigManager';
+import {CUSTOM_SHARE_KEYS, ORIGIN, ICON_SIZE, ALT_TEXT} from './constants';
+
+// This `configManager` will be shared by all AmpAddThis elements on a page, to prevent unnecessary
+// HTTP requests, get accurate analytics, etc.
+const configManager = new ConfigManager();
+
+export function getConfigManager() {
+  return configManager;
+}
+
+class AmpAddThis extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?HTMLIFrameElement} */
+    this.iframe_ = null;
+
+    /** @private {?Promise} */
+    this.iframeLoadPromise_ = null;
+
+    /** @private {string} */
+    this.pubId_ = '';
+
+    /** @private {string} */
+    this.widgetId_ = '';
+
+    /** @private {(?Object<string, string>|undefined)} */
+    this.shareConfig_ = null;
+  }
+
+  /**
+   * @param {boolean=} opt_onLayout
+   * @override
+   */
+  preconnectCallback(opt_onLayout) {
+    const {preconnect} = this;
+    preconnect.url(ORIGIN, opt_onLayout);
+    preconnect.url('https://m.addthis.com', opt_onLayout);
+    preconnect.url('https://m.addthisedge.com', opt_onLayout);
+    preconnect.url('https://cache.addthiscdn.com', opt_onLayout);
+  }
+
+  /**
+   * @return {boolean}
+   * @override
+   */
+  renderOutsideViewport() {
+    return true;
```

Comment:
let's keep normal resource scheduling. I don't see a reason to make an exception for this component.

Reply:
One of the main reasons is this: publishers rely on AddThis tools to provide them with accurate analytics about _visits_ to the pages on which the tools reside. It's fair to say that this is a core use case for our publishers (they like to see data, for example, on things like the number of shares relative to the number of page visits). If the tools don't render unless they are scrolled into view (etc.), then the analytics data provided to publishers will be inaccurate, as not all visits will be registered correctly (since registering the visit using our tools requires at least that the iframe be loaded, which currently happens in`layoutCallback`). Does that make sense?

Reply:
If you'd like to send some analytics ping, the right way to do it is to just emit an image call at the right time. We don't allow iframe based analytics.

I think @zhouyx was working on extension analytics. But something simpler might work here.

Reply:
@jmrog per @cramforce's comment, you can do:

```
buildCallback() {
 ...
 viewer.whenFirstVisible().then( () => {
   const pixel = new thisn.win.Image();
   pixel.src = 'image-based source for your analytics'
 }
}
```

Reply:
Extension analytics doesn't apply here. Extension analytics make sure that the analytics is scoped in the extension element, not about the page. 
I think `#buildCallback()` is a good time to send the analytics ping.

Reply:
Hi all, thanks for the input here. Dropping a pixel during the `buildCallback` is a reasonable option for registering a visit to a page; however, does this mean there is no AMP-compliant way to retrieve data from addthis.com cookies client-side in order to send it with the initial visit request? That will cripple some of the analytics that we currently provide to publishers with our standard tools, so I want to be sure.